parachain-upgrade/src/lib.rs

```diff
@@ -0,0 +1,633 @@
+			let apply_block = vfp.code_upgrade_allowed.ok_or(Error::<T>::ProhibitedByPolkadot)?;
+
+			// When a code upgrade is scheduled, it has to be applied in two
+			// places, synchronized: both polkadot and the individual parachain
+			// have to upgrade on the same relay chain block.
+			//
+			// `notify_polkadot_of_pending_upgrade` notifies polkadot; the `PendingValidationFunction`
+			// storage keeps track locally for the parachain upgrade, which will
+			// be applied later.
+			Self::notify_polkadot_of_pending_upgrade(&validation_function);
+			PendingValidationFunction::put((apply_block, validation_function));
+			Self::deposit_event(Event::ValidationFunctionStored(apply_block));
+		}
+
+		/// Set the current validation function parameters
+		///
+		/// This should be invoked exactly once per block. It will panic at the finalization
+		/// phease if the call was not invoked.
+		///
+		/// The dispatch origin for this call must be `Inherent`
+		///
+		/// As a side effect, this function upgrades the current validation function
+		/// if the appropriate time has come.
+		//
+		// weight data just stolen from Timestamp::set; may be inappropriate
+		#[weight = (0, DispatchClass::Mandatory)]
+		fn set_validation_function_parameters(origin, vfp: ValidationFunctionParams) {
+			ensure_none(origin)?;
+			assert!(!DidUpdateVFPs::exists(), "VFPs must be updated only once in the block");
+
+			// initialization logic: we know that this runs exactly once every block,
+			// which means we can put the initialization logic here to remove the
+			// sequencing problem.
+			if let Ok((apply_block, validation_function)) = PendingValidationFunction::try_get() {
+				if vfp.relay_chain_height >= apply_block {
+					PendingValidationFunction::kill();
+					Self::put_parachain_code(&validation_function);
+					Self::deposit_event(Event::ValidationFunctionApplied(vfp.relay_chain_height));
+				}
+			}
+
+			storage::unhashed::put(VALIDATION_FUNCTION_PARAMS, &vfp);
+			DidUpdateVFPs::put(true);
+			<T::OnValidationFunctionParams as OnValidationFunctionParams>::on_validation_function_params(vfp);
+		}
+
+		fn on_finalize() {
+			assert!(DidUpdateVFPs::take(), "VFPs must be updated once per block");
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Get validation function parameters.
+	fn validation_function_params() -> ValidationFunctionParams {
+		// this storage value is set by cumulus during block validation,
+		// and also by the inherent from this module.
+		storage::unhashed::get(VALIDATION_FUNCTION_PARAMS)
+			.expect("validation function params must be set")
+	}
+
+	/// Get validation function parameters.
+	///
+	/// This will return `None` if this module's inherent has not yet run.
+	/// If it returns `Some(_)`, the validation function params are current for this block.
+	pub fn get_validation_function_params() -> Option<ValidationFunctionParams> {
+		if DidUpdateVFPs::try_get().unwrap_or(false) {
```

Comment:
```suggestion
		if DidUpdateVFPs::get() {
```

`get()` will fallback to `Default::default()` if the value does not exists and `Default` for bool is `false`.

Reply:
I'm aware; I chose this approach to make it more explicit. I'll change it in a minute.